### PR TITLE
ENH: sparse: add count_nonzero method

### PR DESCRIPTION
**Pull request by perimosocordiae**
This is a followup of #3524 by @larsmans.

Tests cover all formats, including non-canonical cases. There are currently two problems I'd like some feedback on:
- <strike>Should I copy the docstring from the `spmatrix` implementation to each of the subclass' implementations? Also, is the docstring appropriate?</strike>
- <strike>Non-canonical matrices currently fall back to the `spmatrix` generic implementation, which performs densification. I did this because it seemed like a `count_nonzero` method shouldn't ever mutate state, and the CSR/CSC `eliminate_zeros` method is in-place. Perhaps this case could be handled by making a copy and eliminating zeros on that?</strike>

### CODE REVIEW

**Comment by perimosocordiae** [comment]
Bump?

File: scipy/sparse/dok.py [hunk 1]
```diff
@@ -115,6 +115,9 @@ def getnnz(self):
         return dict.__len__(self)
     nnz = property(fget=getnnz)
 
+    def count_nonzero(self):
+        return sum(x != 0 for x in self.values())
```
Comment (pv) [inline scipy/sparse/dok.py]: six.itervalues

File: scipy/sparse/base.py [hunk 2]
```diff
@@ -140,6 +140,19 @@ def getmaxprint(self):
...
+        """Number of non-zero entries, equivalent to
+
+        np.count_nonzero(a.toarray())
+
+        Unlike getnnz and the nnz property, which return the number of stored
+        entries (the length of the data attribute), this method counts the
+        actual number of non-zero entries in data.
+        """
+        # Subclasses may implement more efficient versions of this method,
+        # but this acts as a generic fallback in case they do not.
```
Comment (pv) [inline scipy/sparse/base.py]: Do we have subclasses that don't implement this?
If not, maybe prefer to raise NotImplementedError?

File: scipy/sparse/data.py [hunk 3]
```diff
@@ -66,6 +66,11 @@ def conj(self):
     def copy(self):
         return self._with_data(self.data.copy(), copy=True)
 
+    def count_nonzero(self):
+        if self.has_canonical_format:
+            return np.count_nonzero(self.data)
```
Comment (pv) [inline scipy/sparse/data.py]: Maybe should `sum_duplicates`?

File: scipy/sparse/dia.py [hunk 4]
```diff
@@ -151,6 +151,15 @@ def __repr__(self):
...
                  _formats[format][1],))
 
+    def count_nonzero(self):
+        num_rows, num_cols = self.shape
+        offset_inds = np.arange(self.data.shape[1])
+        row = offset_inds - self.offsets[:,None]
+        mask = (row >= 0)
+        mask &= (row < num_rows)
+        mask &= (offset_inds < num_cols)
+        return np.count_nonzero(self.data[mask])
```
Comment (pv) [inline scipy/sparse/dia.py]: Looks like it would be best to add `count_nonzero` tests with non-square matrices (both tall and wide...)

**Comment by pv** [comment]
Now that we more or less decided what to do, maybe push this forward again. LGTM mainly. Is `count_nonzero()` a good name (maybe, following Numpy) --- what did Julia choose?

As per the previous discussions, we opted for canonicalizing when necessary.

The documentation probably should be updated a bit, to indicate `nnz` are *structural nonzeros*, `count_zeros()` are actual nonzeros. These should go to the respective docstrings (for nnz, there's a property docstring?). It's probably OK to just do `count_nonzeros.__doc__ = spmatrix.count_nonzeros.__doc__` to avoid repeating the same docstring.

**Comment by perimosocordiae** [comment]
I think I've addressed all your comments, @pv. The most recent commit also:

- deduplicates docstrings for `getnnz()`, `count_nonzero()`, and the `nnz` property.
- adds the `axis` kwarg to all `getnnz()` methods, with appropriate errors where not implemented.
- uses a separate docstring for the `nnz` property, to avoid mentioning the `axis` kwarg.

File: scipy/sparse/dia.py (hunk 4 above)
Comment (perimosocordiae) [inline scipy/sparse/dia.py]: The default test matrix has shape (3,4), so I added a test using its transpose as well.

**Comment by pv** [comment]
thanks, merged

**Pull request merged**